src/compile/data/optimize.ts

```diff
@@ -108,6 +109,35 @@ function getLeaves(roots: DataFlowNode[]) {
   return leaves;
 }
 
+/**
+ * ParseNodes are not mergeable only if their `this._parse`s have a common key but different values for that key.
+ */
+export function isParseMergeable(x: ParseNode, y: ParseNode) {
+  const xParse = x.parse;
+  const yParse = y.parse;
+  for (const k of keys(xParse)) {
+    if (k in yParse && yParse[k] !== xParse[k]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function mergeParse(node: DataFlowNode) {
+  const parseChildren = node.children.filter((x): x is ParseNode => x instanceof ParseNode);
+
+  if (parseChildren.length > 1) {
+    const mergedParse = parseChildren.pop();
+    const mergeableParseChildren = parseChildren.filter(x => isParseMergeable(x, mergedParse));
+    mergeableParseChildren.forEach(x => {
+      node.removeChild(x);
```

Comment:
This is still removing and adding later, which means it won't preserve the order. Better call `x.remove()` so we can leverage Dom's change in #4038